### PR TITLE
aspsp new endpoints

### PR DESCRIPTION
**Pull request by spetrychenko**

### CODE REVIEW

File: xs2a-adapter-app/src/main/java/de/adorsys/xs2a/adapter/config/RestConfiguration.java [hunk 1]
```diff
@@ -45,4 +48,10 @@ AspspRepository aspspRepository() {
     AspspSearchService aspspSearchService(AspspRepository aspspRepository) {
         return new AspspSearchServiceImpl(aspspRepository);
     }
+
+    @Profile("dev")
+    @Bean
+    AspspModifyRepository aspspModifyRepository(){
```
Comment (artemsnisarenko) [inline xs2a-adapter-app/src/main/java/de/adorsys/xs2a/adapter/config/RestConfiguration.java]: Whitespace is missing.

File: xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/AspspSearchServiceImpl.java [hunk 2]
```diff
@@ -32,7 +32,9 @@ public AspspSearchServiceImpl(AspspRepository luceneAspspRepository) {
             throw new IbanException(e.getMessage());
         }
         if (bankCode == null) {
-            throw new IbanException("Failed to extract the bank code from the iban");
+            String errorMessage = "Failed to extract the bank code from the iban=" + iban;
```
Comment (artemsnisarenko) [inline xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/AspspSearchServiceImpl.java]: https://stackoverflow.com/questions/6639963/why-is-log-and-throw-considered-an-anti-pattern

File: xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepository.java [hunk 3]
```diff
@@ -42,15 +45,61 @@ public void save(Aspsp aspsp) {
         }
     }
 
+    @Override
+    public Aspsp create(Aspsp aspsp) {
+        Optional<Aspsp> optional = findById(aspsp.getId());
+        if (optional.isPresent()){
```
Comment (artemsnisarenko) [inline xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepository.java]: Whitespaces...

File: xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepository.java [hunk 4]
```diff
@@ -42,15 +45,61 @@ public void save(Aspsp aspsp) {
         }
     }
 
+    @Override
+    public Aspsp create(Aspsp aspsp) {
+        Optional<Aspsp> optional = findById(aspsp.getId());
```
Comment (artemsnisarenko) [inline xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepository.java]: Calling `findById` with a null will result in NPE

File: xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepository.java [hunk 5]
```diff
@@ -42,15 +45,61 @@ public void save(Aspsp aspsp) {
...
+            throw new DuplicationAspspException("Aspsp with id="+aspsp.getId()+" already exists");
+        }
+        save(aspsp);
+        return aspsp;
+    }
+
+    @Override
+    public void update(Aspsp aspsp) {
+        findById(aspsp.getId())
+            .orElseThrow(()->new AspspRegistrationNotFoundException("Aspsp with id="+aspsp.getId()+" was not found"));
```
Comment (artemsnisarenko) [inline xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepository.java]: again whitespaces

File: xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepository.java [hunk 6]
```diff
@@ -113,28 +162,28 @@ private Query getBicQuery(String bic) {
             TopDocs topDocs = indexSearcher.searchAfter(afterDoc, query, size);
             ScoreDoc[] scoreDocs = topDocs.scoreDocs;
             return Arrays.stream(scoreDocs)
-                .map(this::getDocumentAsAspsp)
-                .filter(Optional::isPresent)
-                .map(Optional::get)
-                .collect(toList());
+                       .map(this::getDocumentAsAspsp)
```
Comment (artemsnisarenko) [inline xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepository.java]: please revert

File: xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/exception/DuplicationAspspException.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package de.adorsys.xs2a.adapter.registry.exception;
+
+public class DuplicationAspspException extends RegistryIOException {
```
Comment (artemsnisarenko) [inline xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/exception/DuplicationAspspException.java]: It has nothing to do with `IOException` and thus should not extend `RegistryIOException`. Also a handler for rest is necessary.

File: xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/exception/RegistryIOException.java [hunk 8]
```diff
@@ -5,6 +5,13 @@
 
 public class RegistryIOException extends RuntimeException {
 
+    public RegistryIOException() {
```
Comment (artemsnisarenko) [inline xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/exception/RegistryIOException.java]: should be reverted

File: xs2a-adapter-aspsp-registry/src/test/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepositoryTest.java [hunk 9]
```diff
@@ -14,6 +16,57 @@
...
+        luceneAspspRepository.save(aspsp);
+
+        aspsp.setId("1");
+        luceneAspspRepository.create(aspsp);
+
+        luceneAspspRepository.create(aspsp);
+    }
+
+    @Test
+    public void delete() {
```
Comment (artemsnisarenko) [inline xs2a-adapter-aspsp-registry/src/test/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepositoryTest.java]: This and the following test don't assert anything.

File: xs2a-adapter-rest-server/src/main/java/de/adorsys/xs2a/adapter/controller/AspspModifyController.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+    @PostMapping(AspspService.V1_APSPS)
+    ResponseEntity<AspspTO> addAspsp(@RequestBody AspspTO to) {
+        Aspsp aspsp = aspspModifyRepository.create(aspspMapper.toAspsp(to));
+        String uri = V1_ASPSP_BY_ID.replace(ASPSP_ID, aspsp.getId());
+        return ResponseEntity.created(URI.create(uri)).body(aspspMapper.toAspspTO(aspsp));
+    }
+
+    @PutMapping(AspspService.V1_APSPS)
+    ResponseEntity<AspspTO> updateAspsp(@RequestBody AspspTO aspsp){
```
Comment (artemsnisarenko) [inline xs2a-adapter-rest-server/src/main/java/de/adorsys/xs2a/adapter/controller/AspspModifyController.java]: whitespace problems in this file as well

File: xs2a-adapter-rest-server/src/main/java/de/adorsys/xs2a/adapter/controller/AspspModifyController.java [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    static final String V1_ASPSP_BY_ID = AspspService.V1_APSPS + "/" + ASPSP_ID;
+    private final AspspModifyRepository aspspModifyRepository;
+    private final AspspMapper aspspMapper = Mappers.getMapper(AspspMapper.class);
+
+    public AspspModifyController(AspspModifyRepository aspspModifyRepository) {
+        this.aspspModifyRepository = aspspModifyRepository;
+    }
+
+    @PostMapping(AspspService.V1_APSPS)
+    ResponseEntity<AspspTO> addAspsp(@RequestBody AspspTO to) {
```
Comment (artemsnisarenko) [inline xs2a-adapter-rest-server/src/main/java/de/adorsys/xs2a/adapter/controller/AspspModifyController.java]: inconsistent method names

File: xs2a-adapter-rest-server/src/main/java/de/adorsys/xs2a/adapter/controller/AspspModifyController.java [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    @PutMapping(AspspService.V1_APSPS)
+    ResponseEntity<AspspTO> updateAspsp(@RequestBody AspspTO aspsp){
+        aspspModifyRepository.update(aspspMapper.toAspsp(aspsp));
+        return ResponseEntity.ok(aspsp);
+    }
+
+    @DeleteMapping(V1_ASPSP_BY_ID)
+    ResponseEntity<Void> delete(@PathVariable("aspspId") String aspspId){
```
Comment (artemsnisarenko) [inline xs2a-adapter-rest-server/src/main/java/de/adorsys/xs2a/adapter/controller/AspspModifyController.java]: inconsistent method names again

File: xs2a-adapter-rest-server/src/main/java/de/adorsys/xs2a/adapter/controller/AspspService.java [hunk 13]
```diff
@@ -19,12 +19,16 @@
...
 import org.springframework.http.ResponseEntity;
 import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
 import org.springframework.web.bind.annotation.RequestParam;
 
 import java.util.List;
 
 public interface AspspService {
-    @RequestMapping("/v1/aspsps")
+
```
Comment (artemsnisarenko) [inline xs2a-adapter-rest-server/src/main/java/de/adorsys/xs2a/adapter/controller/AspspService.java]: I think this code should be in `AspspController` and `AspspService` is removed. It's confusing in it's implements relation to controller compared to `AspspSearchService` and `AspspModifyService`.

File: xs2a-adapter-service-api/src/main/java/de/adorsys/xs2a/adapter/service/AspspModifyRepository.java [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package de.adorsys.xs2a.adapter.service;
+
+import de.adorsys.xs2a.adapter.service.model.Aspsp;
+
+public interface AspspModifyRepository {
```
Comment (artemsnisarenko) [inline xs2a-adapter-service-api/src/main/java/de/adorsys/xs2a/adapter/service/AspspModifyRepository.java]: These "modify" infixes look horrible to me. Have you considered any alternatives?
Perhaps `AspspRepository` should be `AspspReadonlyRepository` and `AspspModifyRepository` extend it and be named `AspspRepository`.
Also please use standard method names https://docs.spring.io/spring-data/commons/docs/current/api/org/springframework/data/repository/CrudRepository.html

File: xs2a-adapter-rest-server/src/main/java/de/adorsys/xs2a/adapter/controller/AspspService.java (hunk 13 above)
Comment (spetrychenko) [reply xs2a-adapter-rest-server/src/main/java/de/adorsys/xs2a/adapter/controller/AspspService.java]: I do not agree. We need such an interface for Feign remote client. I will rename it to AspspApi.

File: xs2a-adapter-service-api/src/main/java/de/adorsys/xs2a/adapter/service/AspspModifyRepository.java (hunk 14 above)
Comment (spetrychenko) [reply xs2a-adapter-service-api/src/main/java/de/adorsys/xs2a/adapter/service/AspspModifyRepository.java]: renamed AspspRepository to AspspReadonlyRepository

Comment (artemsnisarenko) [reply xs2a-adapter-service-api/src/main/java/de/adorsys/xs2a/adapter/service/AspspModifyRepository.java]: In my opinion this doesn't cut it.

File: xs2a-adapter-rest-server/src/main/java/de/adorsys/xs2a/adapter/controller/AspspService.java (hunk 13 above)
Comment (artemsnisarenko) [reply xs2a-adapter-rest-server/src/main/java/de/adorsys/xs2a/adapter/controller/AspspService.java]: In this case why the controller you've added doesn't have an interface?

File: xs2a-adapter-aspsp-registry/src/test/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepositoryTest.java [hunk 15]
```diff
@@ -136,8 +161,8 @@ public void findAllWithPagination() {
         Aspsp aspsp3 = new Aspsp();
         luceneAspspRepository.save(aspsp3);
 
-        List<Aspsp> firstPage = luceneAspspRepository.findAll(2);
-        assertThat(firstPage).hasSize(2);
+        List<Aspsp> firstPage = luceneAspspRepository.findAll(3);
```
Comment (artemsnisarenko) [inline xs2a-adapter-aspsp-registry/src/test/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepositoryTest.java]: Why is this changed? The tests inserts 3 records and then verifies that with page size = 2 there are 2 records returned on the first page and one on the second.

File: xs2a-adapter-rest-server/src/main/java/de/adorsys/xs2a/adapter/controller/AspspService.java (hunk 13 above)
Comment (spetrychenko) [reply xs2a-adapter-rest-server/src/main/java/de/adorsys/xs2a/adapter/controller/AspspService.java]: We do not want to expose modifying operations. They will be used only with `dev` profile

File: xs2a-adapter-aspsp-registry/src/test/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepositoryTest.java (hunk 15 above)
Comment (spetrychenko) [reply xs2a-adapter-aspsp-registry/src/test/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepositoryTest.java]: sorry, my fault. reverted back

File: xs2a-adapter-service-api/src/main/java/de/adorsys/xs2a/adapter/service/AspspModifyRepository.java (hunk 14 above)
Comment (spetrychenko) [reply xs2a-adapter-service-api/src/main/java/de/adorsys/xs2a/adapter/service/AspspModifyRepository.java]: please suggest a better name.

File: xs2a-adapter-app/src/main/java/de/adorsys/xs2a/adapter/config/RestConfiguration.java (hunk 1 above)
Comment (spetrychenko) [reply xs2a-adapter-app/src/main/java/de/adorsys/xs2a/adapter/config/RestConfiguration.java]: thanks, fixed

File: xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/AspspSearchServiceImpl.java (hunk 2 above)
Comment (spetrychenko) [reply xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/AspspSearchServiceImpl.java]: thanks, fixed

File: xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepository.java (hunk 3 above)
Comment (spetrychenko) [reply xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepository.java]: thanks, fixed

File: xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepository.java (hunk 4 above)
Comment (spetrychenko) [reply xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepository.java]: thanks, fixed

File: xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepository.java (hunk 5 above)
Comment (spetrychenko) [reply xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepository.java]: thanks, fixed

File: xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepository.java (hunk 6 above)
Comment (spetrychenko) [reply xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepository.java]: thanks, fixed

File: xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/exception/DuplicationAspspException.java (hunk 7 above)
Comment (spetrychenko) [reply xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/exception/DuplicationAspspException.java]: thanks, fixed

File: xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/exception/RegistryIOException.java (hunk 8 above)
Comment (spetrychenko) [reply xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/exception/RegistryIOException.java]: thanks, fixed

File: xs2a-adapter-aspsp-registry/src/test/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepositoryTest.java (hunk 9 above)
Comment (spetrychenko) [reply xs2a-adapter-aspsp-registry/src/test/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepositoryTest.java]: thanks, fixed

File: xs2a-adapter-rest-server/src/main/java/de/adorsys/xs2a/adapter/controller/AspspModifyController.java (hunk 10 above)
Comment (spetrychenko) [reply xs2a-adapter-rest-server/src/main/java/de/adorsys/xs2a/adapter/controller/AspspModifyController.java]: thanks, fixed

File: xs2a-adapter-rest-server/src/main/java/de/adorsys/xs2a/adapter/controller/AspspModifyController.java (hunk 11 above)
Comment (spetrychenko) [reply xs2a-adapter-rest-server/src/main/java/de/adorsys/xs2a/adapter/controller/AspspModifyController.java]: thanks, fixed

File: xs2a-adapter-rest-server/src/main/java/de/adorsys/xs2a/adapter/controller/AspspModifyController.java (hunk 12 above)
Comment (spetrychenko) [reply xs2a-adapter-rest-server/src/main/java/de/adorsys/xs2a/adapter/controller/AspspModifyController.java]: thanks, fixed
